libgalois/include/katana/ArrowRandomAccessBuilder.h

```diff
@@ -10,182 +11,186 @@ namespace katana {
+    return boost::atomic_ref<uint8_t>(*ptr);
+  }
 };
 
 template <typename ArrowType>
-struct ArrowTypeConfig;
+struct RandomBuilderTypeConfig;
 
-#define NULLABLE(ValueType, StorageType, ArrowType)                            \
+#define VECTOR_BACKED_BUILDER(ValueType, StorageType, ArrowType)               \
   template <>                                                                  \
-  struct ArrowTypeConfig<ArrowType> {                                          \
+  struct RandomBuilderTypeConfig<ArrowType> {                                  \
     using RandomBuilderType =                                                  \
-        NullableBuilder<ValueType, StorageType, ArrowType>;                    \
+        VectorBackedBuilder<ValueType, StorageType, ArrowType>;                \
   }
 
-NULLABLE(int8_t, int8_t, arrow::Int8Type);
-NULLABLE(uint8_t, uint8_t, arrow::UInt8Type);
-NULLABLE(int16_t, int16_t, arrow::Int16Type);
-NULLABLE(uint16_t, uint16_t, arrow::UInt16Type);
-NULLABLE(int32_t, int32_t, arrow::Int32Type);
-NULLABLE(uint32_t, uint32_t, arrow::UInt32Type);
-NULLABLE(int64_t, int64_t, arrow::Int64Type);
-NULLABLE(uint64_t, uint64_t, arrow::UInt64Type);
-NULLABLE(float, float, arrow::FloatType);
-NULLABLE(double, double, arrow::DoubleType);
-NULLABLE(bool, uint8_t, arrow::BooleanType);
-NULLABLE(std::string, std::string, arrow::StringType);
-NULLABLE(std::string, std::string, arrow::LargeStringType);
-
-#undef NULLABLE
+#define IN_PLACE_BUILDER(ArrowType)                                            \
+  template <>                                                                  \
+  struct RandomBuilderTypeConfig<ArrowType> {                                  \
+    using RandomBuilderType = InPlaceBuilder<ArrowType>;                       \
+  }
+
+// arrow::BooleanArray is bit-compacted, promote to byte for thread-safety
+VECTOR_BACKED_BUILDER(bool, uint8_t, arrow::BooleanType);
+
+// Intermediate storage is mandatory for non-PODs, as Data size is unknown
+VECTOR_BACKED_BUILDER(std::string, std::string, arrow::StringType);
+VECTOR_BACKED_BUILDER(std::string, std::string, arrow::LargeStringType);
+
+// TODO(daniel) these should be fine as InPlace
+// For some reason, using InPlace makes the partitioner
+// unable to partition Types, stored as uint8_t
+// Marking anything smaller than 32-bit as Vector for now
+VECTOR_BACKED_BUILDER(int8_t, int8_t, arrow::Int8Type);
+VECTOR_BACKED_BUILDER(uint8_t, uint8_t, arrow::UInt8Type);
+VECTOR_BACKED_BUILDER(int16_t, int16_t, arrow::Int16Type);
+VECTOR_BACKED_BUILDER(uint16_t, uint16_t, arrow::UInt16Type);
+
+VECTOR_BACKED_BUILDER(int32_t, int32_t, arrow::Int32Type);
+VECTOR_BACKED_BUILDER(uint32_t, uint32_t, arrow::UInt32Type);
+VECTOR_BACKED_BUILDER(int64_t, int64_t, arrow::Int64Type);
+VECTOR_BACKED_BUILDER(uint64_t, uint64_t, arrow::UInt64Type);
+VECTOR_BACKED_BUILDER(float, float, arrow::FloatType);
+VECTOR_BACKED_BUILDER(double, double, arrow::DoubleType);
```

Comment:
The comment above suggests that in place builder should work for these values?

Reply:
Indeed it should (or at least I believed it should). Any fixed-width type is eligible except bool (since I didn't write an overload to use atomic_ref on both the null bitmask and the value itself, and that would be fundamentally kinda complicated, because reference types)